libfastpm/factors.c

```diff
@@ -230,37 +235,83 @@ void fastpm_kick_init(FastPMKickFactor * kick, FastPMSolver * fastpm, double ai,
     FastPMCosmology * c = fastpm->cosmology;
     kick->forcemode = fastpm->config->FORCE_TYPE;
 
-    double OmegaM = c->OmegaM;
-    double Om143 = pow(OmegaA(ac, c), 1.0/143.0);
-    double growth1 = GrowthFactor(ac, c);
-
-    kick->q1 = growth1;
-    kick->q2 = growth1*growth1*(1.0 + 7.0/3.0*Om143);
+    FastPMGrowthInfo gi_i;
+    FastPMGrowthInfo gi_c;
+    FastPMGrowthInfo gi_e;
+
+    fastpm_growth_info_init(&gi_i, ai, c);
+    fastpm_growth_info_init(&gi_c, ac, c);
+
+    double E_i = HubbleEa(ai, c);
+    double E_c = HubbleEa(ac, c);
+
+    double D1_i = gi_i.D1;
+    double D2_i = gi_i.D2;
+    double f1_i = gi_i.f1;
+    double f2_i = gi_i.f2;
+
+    double D1_c = gi_c.D1;
+    double D2_c = gi_c.D2;
+
+    double Omega_m0 = c->Omega_m;
+    double Omega_mc = Omega_m(ac, c);   // assumes all ncdm is matter like
+
+    kick->q1 = D1_c;
+    switch (c->growth_mode){
+        case FASTPM_GROWTH_MODE_LCDM:
+            kick->q2 = D1_c*D1_c * (1.0 + 7.0/3.0 * pow(Omega_mc, 1.0/143.0));
+        break;
+        case FASTPM_GROWTH_MODE_ODE:
+            kick->q2 = D1_c*D1_c * (1 - D1_c*D1_c/D2_c);  // FIXME: This matches up with old code mathematically, but review this!
```

Comment:
A way to always check this is compute both Lcdm values, and raise an error if they are too far off.

Reply:
They're not meant to be equal for a radiation cosmology, but they should be approx equal in the Lambda+CDM only case. So it would be a bit messy to check. The question is if this is the correct way to generalize q_2. But anyway, this is only used for the COLA force implementation, so maybe not a huge deal?